src/hypergol/tensorflow_model_manager.py

```diff
@@ -0,0 +1,146 @@
+        if withLogging:
+            with self.trainingSummaryWriter.as_default():
+                tf.summary.scalar(name='Loss', data=loss, step=self.globalStep)
+                if withMetadata and self.globalStep > 0:
+                    tf.summary.trace_export(name=f'{self.model.__class__.__name__}{self.globalStep}', step=self.globalStep, profiler_outdir=f'{self.tensorboardPath}/trainGraph')
+        self.globalStep += 1
+        return loss
+
+    def evaluate(self, withLogging, withMetadata):
+        """runs an evaluation step for the model
+
+        Parameters
+        ----------
+        withLogging: bool
+            produce tensorflow logging for step
+        withMetadata: bool
+            log tensorflow graph metadata for step
+        """
+        batch = next(self.batchReader)
+        if withMetadata and self.globalStep > 0:
+            tf.summary.trace_on(graph=True, profiler=False)
+        outputs, loss, metrics = self.model.evaluate(inputs=batch['inputs'], targets=batch['targets'])
+        if withLogging:
+            with self.evaluationSummaryWriter.as_default():
+                tf.summary.scalar(name='Loss', data=loss, step=self.globalStep)
+                for scalarName, scalarMetric in metrics.get_scalar_metrics().items():
+                    tf.summary.scalar(name=f'Evaluation/{scalarName}', data=scalarMetric, step=self.globalStep)
+                for histogramName, histogramMetric in metrics.get_histogram_metrics().items():
+                    tf.summary.histogram(name=f'Evaluation/{histogramName}', data=histogramMetric, step=self.globalStep)
+                if withMetadata and self.globalStep > 0:
+                    tf.summary.trace_export(name=f'{self.model.__class__.__name__}{self.globalStep}', step=self.globalStep, profiler_outdir=f'{self.tensorboardPath}/evaluateGraph')
+        if self.outputSaver is not None:
+            self.outputSaver.save_outputs(batch=batch, outputs=outputs, globalStep=self.globalStep)
+        return outputs, loss, metrics
+
+    def checkpoint(self):
+        """checkpoint the model"""
+        self.model.checkpoint(path=self.checkpointFolder)
+
+    def run(self, stepCount, evaluationSteps, tensorboardSteps, metadataSteps, trainingSteps=None):
+        """runs a training schedule
+
+        Parameters
+        ----------
+        stepCount: int
+            num total steps in schedule
+        evaluationSteps: list(int)
+            which steps to perform evaluation
+        tensorboardSteps: list(int)
+            which steps to log metrics to tensorboard
+        metadataSteps: list(int)
+            which steps to log metadata to tensorboard
+        trainingSteps: list(int)
+            which steps to train model
+        """
+        if trainingSteps is None:
+            trainingSteps = range(stepCount)
+        self.initialize()
+        for k in tqdm(range(stepCount)):
```

Comment:
Don't quite follow whats going on here but feel like 'k' might not be an ideal variable name in this context?